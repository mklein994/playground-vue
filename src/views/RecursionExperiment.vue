<script setup lang="ts">
import { computed, ref } from "vue";

import {
  dataGet,
  dataGetObject,
  dataSet,
  dataSetObject,
  type RecursiveMap,
} from "@/helpers/dataSet";

const props = withDefaults(
  defineProps<{
    source?: string[];
    split?: (word: string) => string[];
    search?: string;
  }>(),
  {
    source: () => Object.keys(import.meta.glob("/src/**")),
    split: (word: string) => word.match(/\/[^/]+/g) ?? [],
    search: "/src/App.vue",
  }
);

const sourceInput = ref(props.search);
const sourceTree = props.source.reduce(
  (all, one) => dataSet(all, props.split(one), one),
  new Map() as RecursiveMap
);

const sourceFromTree = computed(
  () => dataGet(sourceTree, props.split(sourceInput.value)) ?? "(undefined)"
);
const sourceFromObject = computed(
  () =>
    dataGetObject(sourceObject, props.split(sourceInput.value)) ?? "(undefined)"
);

const sourceObject = props.source.reduce(
  (all, one) => dataSetObject(all, props.split(one), one),
  {} as Record<string, unknown>
);
</script>

<template>
  <div class="recursion">
    <input v-model="sourceInput" type="text" />
    <output>{{ sourceFromTree }}</output>
    <output>{{ sourceFromObject }}</output>
  </div>
  <ul>
    <li v-for="file of source" :key="file">{{ file }}</li>
  </ul>
  <pre>{{ sourceTree }}</pre>
  <pre>{{ sourceObject }}</pre>
</template>

<style scoped>
.recursion {
  display: grid;
  max-width: max-content;
}
</style>
