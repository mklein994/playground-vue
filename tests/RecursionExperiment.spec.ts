import { flushPromises, mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";

import RecursionExperiment from "@/views/RecursionExperiment.vue";

describe("RecursionExperiment", () => {
  // This test can't be run concurrently, since it sometimes causes the
  // snapshot to be discarded, marking it as "obsolete". I'm not sure why.
  it("renders correctly", async () => {
    const wrapper = mount(RecursionExperiment, {
      props: {
        source: [
          "store.produce.apple",
          "store.produce.banana",
          "store.produce.cherry",
          "person.male.dad",
          "person.male.son",
          "person.female.mom",
          "person.female.daughter",
        ],
        split: (word: string) => word.split("."),
        search: "store.produce.apple",
      },
    });
    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("handles an empty source", async () => {
    const wrapper = mount(RecursionExperiment, {
      props: {
        source: ["foo.bar"],
        search: "baz",
        split: (word: string) => word.split("."),
      },
    });
    await flushPromises();

    expect(
      wrapper
        .findAll("output")
        .map((x) => x.text())
        .every((x) => x === "(undefined)"),
      "not every output contains '(undefined)'"
    ).toBe(true);
  });
});
