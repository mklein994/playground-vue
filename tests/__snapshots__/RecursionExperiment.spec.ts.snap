// Vitest Snapshot v1

exports[`RecursionExperiment 1`] = `
"<div class=\\"recursion\\" data-v-54f64b35=\\"\\"><input type=\\"text\\" data-v-54f64b35=\\"\\" /><output data-v-54f64b35=\\"\\">/src/App.vue</output><output data-v-54f64b35=\\"\\">/src/App.vue</output></div>
<ul data-v-54f64b35=\\"\\">
  <li data-v-54f64b35=\\"\\">/src/App.vue</li>
  <li data-v-54f64b35=\\"\\">/src/HomeView.vue</li>
  <li data-v-54f64b35=\\"\\">/src/NavigationFragment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/NotFound.vue</li>
  <li data-v-54f64b35=\\"\\">/src/RouteInfo.vue</li>
  <li data-v-54f64b35=\\"\\">/src/env.d.ts</li>
  <li data-v-54f64b35=\\"\\">/src/fonts.css</li>
  <li data-v-54f64b35=\\"\\">/src/main.ts</li>
  <li data-v-54f64b35=\\"\\">/src/routes.ts</li>
  <li data-v-54f64b35=\\"\\">/src/style.css</li>
  <li data-v-54f64b35=\\"\\">/src/tailwind.css</li>
  <li data-v-54f64b35=\\"\\">/src/assets/AndadaPro-VariableFont_wght.woff2</li>
  <li data-v-54f64b35=\\"\\">/src/assets/Recursive.woff2</li>
  <li data-v-54f64b35=\\"\\">/src/components/CustomElement.ce.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/CustomRenderFunction.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/CustomRenderFunctionSetup.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/EventCallbacksHelper.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/ForAwaitList.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/SunriseSunset.vue</li>
  <li data-v-54f64b35=\\"\\">/src/use/dataSet.ts</li>
  <li data-v-54f64b35=\\"\\">/src/use/smokeTest.ts</li>
  <li data-v-54f64b35=\\"\\">/src/views/AboutBrowser.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/AccentColor.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/ConsoleTrace.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/CounterExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/CssLayers.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/CssModules.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/CssVBind.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/CssVariables.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/CustomElement.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/DialogElement.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/DynamicScopedStyles.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/EventCallbacks.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/FooExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/ForAwait.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/GridBorders.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/GridExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/HeroiconsExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/InputFileCapture.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/MultiLineSwitchStatement.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/MultipleCheckboxes.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/NthChild.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/NumberFieldBinding.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/OptionalChaining.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/RecursionExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/RenderFunction.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/SassExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/SelectStyling.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/ServiceWorkerExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/StyleInheritance.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/TaggedTemplates.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/TailwindColors.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/ThrowError.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/TimeElement.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/VariableFonts.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/WasmExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/WorkerExperiment.vue</li>
  <li data-v-54f64b35=\\"\\">/src/views/XorOperation.vue</li>
  <li data-v-54f64b35=\\"\\">/src/workers/my-first-service-worker.ts</li>
  <li data-v-54f64b35=\\"\\">/src/workers/my-first-worker.ts</li>
  <li data-v-54f64b35=\\"\\">/src/components/grid-borders/GridPicker.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/style-inheritance/ChildFour.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/style-inheritance/ChildOne.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/style-inheritance/ChildThree.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/style-inheritance/ChildTwo.vue</li>
  <li data-v-54f64b35=\\"\\">/src/components/variable-fonts/VariableFontAxis.vue</li>
</ul>
<pre data-v-54f64b35=\\"\\">{
  \\"Map(1)\\": {
    \\"/src =>\\": {
      \\"Map(16)\\": {
        \\"/App.vue =>\\": \\"/src/App.vue\\",
        \\"/HomeView.vue =>\\": \\"/src/HomeView.vue\\",
        \\"/NavigationFragment.vue =>\\": \\"/src/NavigationFragment.vue\\",
        \\"/NotFound.vue =>\\": \\"/src/NotFound.vue\\",
        \\"/RouteInfo.vue =>\\": \\"/src/RouteInfo.vue\\",
        \\"/env.d.ts =>\\": \\"/src/env.d.ts\\",
        \\"/fonts.css =>\\": \\"/src/fonts.css\\",
        \\"/main.ts =>\\": \\"/src/main.ts\\",
        \\"/routes.ts =>\\": \\"/src/routes.ts\\",
        \\"/style.css =>\\": \\"/src/style.css\\",
        \\"/tailwind.css =>\\": \\"/src/tailwind.css\\",
        \\"/assets =>\\": {
          \\"Map(2)\\": {
            \\"/AndadaPro-VariableFont_wght.woff2 =>\\": \\"/src/assets/AndadaPro-VariableFont_wght.woff2\\",
            \\"/Recursive.woff2 =>\\": \\"/src/assets/Recursive.woff2\\"
          }
        },
        \\"/components =>\\": {
          \\"Map(9)\\": {
            \\"/CustomElement.ce.vue =>\\": \\"/src/components/CustomElement.ce.vue\\",
            \\"/CustomRenderFunction.vue =>\\": \\"/src/components/CustomRenderFunction.vue\\",
            \\"/CustomRenderFunctionSetup.vue =>\\": \\"/src/components/CustomRenderFunctionSetup.vue\\",
            \\"/EventCallbacksHelper.vue =>\\": \\"/src/components/EventCallbacksHelper.vue\\",
            \\"/ForAwaitList.vue =>\\": \\"/src/components/ForAwaitList.vue\\",
            \\"/SunriseSunset.vue =>\\": \\"/src/components/SunriseSunset.vue\\",
            \\"/grid-borders =>\\": {
              \\"Map(1)\\": {
                \\"/GridPicker.vue =>\\": \\"/src/components/grid-borders/GridPicker.vue\\"
              }
            },
            \\"/style-inheritance =>\\": {
              \\"Map(4)\\": {
                \\"/ChildFour.vue =>\\": \\"/src/components/style-inheritance/ChildFour.vue\\",
                \\"/ChildOne.vue =>\\": \\"/src/components/style-inheritance/ChildOne.vue\\",
                \\"/ChildThree.vue =>\\": \\"/src/components/style-inheritance/ChildThree.vue\\",
                \\"/ChildTwo.vue =>\\": \\"/src/components/style-inheritance/ChildTwo.vue\\"
              }
            },
            \\"/variable-fonts =>\\": {
              \\"Map(1)\\": {
                \\"/VariableFontAxis.vue =>\\": \\"/src/components/variable-fonts/VariableFontAxis.vue\\"
              }
            }
          }
        },
        \\"/use =>\\": {
          \\"Map(2)\\": {
            \\"/dataSet.ts =>\\": \\"/src/use/dataSet.ts\\",
            \\"/smokeTest.ts =>\\": \\"/src/use/smokeTest.ts\\"
          }
        },
        \\"/views =>\\": {
          \\"Map(37)\\": {
            \\"/AboutBrowser.vue =>\\": \\"/src/views/AboutBrowser.vue\\",
            \\"/AccentColor.vue =>\\": \\"/src/views/AccentColor.vue\\",
            \\"/ConsoleTrace.vue =>\\": \\"/src/views/ConsoleTrace.vue\\",
            \\"/CounterExperiment.vue =>\\": \\"/src/views/CounterExperiment.vue\\",
            \\"/CssLayers.vue =>\\": \\"/src/views/CssLayers.vue\\",
            \\"/CssModules.vue =>\\": \\"/src/views/CssModules.vue\\",
            \\"/CssVBind.vue =>\\": \\"/src/views/CssVBind.vue\\",
            \\"/CssVariables.vue =>\\": \\"/src/views/CssVariables.vue\\",
            \\"/CustomElement.vue =>\\": \\"/src/views/CustomElement.vue\\",
            \\"/DialogElement.vue =>\\": \\"/src/views/DialogElement.vue\\",
            \\"/DynamicScopedStyles.vue =>\\": \\"/src/views/DynamicScopedStyles.vue\\",
            \\"/EventCallbacks.vue =>\\": \\"/src/views/EventCallbacks.vue\\",
            \\"/FooExperiment.vue =>\\": \\"/src/views/FooExperiment.vue\\",
            \\"/ForAwait.vue =>\\": \\"/src/views/ForAwait.vue\\",
            \\"/GridBorders.vue =>\\": \\"/src/views/GridBorders.vue\\",
            \\"/GridExperiment.vue =>\\": \\"/src/views/GridExperiment.vue\\",
            \\"/HeroiconsExperiment.vue =>\\": \\"/src/views/HeroiconsExperiment.vue\\",
            \\"/InputFileCapture.vue =>\\": \\"/src/views/InputFileCapture.vue\\",
            \\"/MultiLineSwitchStatement.vue =>\\": \\"/src/views/MultiLineSwitchStatement.vue\\",
            \\"/MultipleCheckboxes.vue =>\\": \\"/src/views/MultipleCheckboxes.vue\\",
            \\"/NthChild.vue =>\\": \\"/src/views/NthChild.vue\\",
            \\"/NumberFieldBinding.vue =>\\": \\"/src/views/NumberFieldBinding.vue\\",
            \\"/OptionalChaining.vue =>\\": \\"/src/views/OptionalChaining.vue\\",
            \\"/RecursionExperiment.vue =>\\": \\"/src/views/RecursionExperiment.vue\\",
            \\"/RenderFunction.vue =>\\": \\"/src/views/RenderFunction.vue\\",
            \\"/SassExperiment.vue =>\\": \\"/src/views/SassExperiment.vue\\",
            \\"/SelectStyling.vue =>\\": \\"/src/views/SelectStyling.vue\\",
            \\"/ServiceWorkerExperiment.vue =>\\": \\"/src/views/ServiceWorkerExperiment.vue\\",
            \\"/StyleInheritance.vue =>\\": \\"/src/views/StyleInheritance.vue\\",
            \\"/TaggedTemplates.vue =>\\": \\"/src/views/TaggedTemplates.vue\\",
            \\"/TailwindColors.vue =>\\": \\"/src/views/TailwindColors.vue\\",
            \\"/ThrowError.vue =>\\": \\"/src/views/ThrowError.vue\\",
            \\"/TimeElement.vue =>\\": \\"/src/views/TimeElement.vue\\",
            \\"/VariableFonts.vue =>\\": \\"/src/views/VariableFonts.vue\\",
            \\"/WasmExperiment.vue =>\\": \\"/src/views/WasmExperiment.vue\\",
            \\"/WorkerExperiment.vue =>\\": \\"/src/views/WorkerExperiment.vue\\",
            \\"/XorOperation.vue =>\\": \\"/src/views/XorOperation.vue\\"
          }
        },
        \\"/workers =>\\": {
          \\"Map(2)\\": {
            \\"/my-first-service-worker.ts =>\\": \\"/src/workers/my-first-service-worker.ts\\",
            \\"/my-first-worker.ts =>\\": \\"/src/workers/my-first-worker.ts\\"
          }
        }
      }
    }
  }
}</pre>
<pre data-v-54f64b35=\\"\\">{
  \\"/src\\": {
    \\"/App.vue\\": \\"/src/App.vue\\",
    \\"/HomeView.vue\\": \\"/src/HomeView.vue\\",
    \\"/NavigationFragment.vue\\": \\"/src/NavigationFragment.vue\\",
    \\"/NotFound.vue\\": \\"/src/NotFound.vue\\",
    \\"/RouteInfo.vue\\": \\"/src/RouteInfo.vue\\",
    \\"/env.d.ts\\": \\"/src/env.d.ts\\",
    \\"/fonts.css\\": \\"/src/fonts.css\\",
    \\"/main.ts\\": \\"/src/main.ts\\",
    \\"/routes.ts\\": \\"/src/routes.ts\\",
    \\"/style.css\\": \\"/src/style.css\\",
    \\"/tailwind.css\\": \\"/src/tailwind.css\\",
    \\"/assets\\": {
      \\"/AndadaPro-VariableFont_wght.woff2\\": \\"/src/assets/AndadaPro-VariableFont_wght.woff2\\",
      \\"/Recursive.woff2\\": \\"/src/assets/Recursive.woff2\\"
    },
    \\"/components\\": {
      \\"/CustomElement.ce.vue\\": \\"/src/components/CustomElement.ce.vue\\",
      \\"/CustomRenderFunction.vue\\": \\"/src/components/CustomRenderFunction.vue\\",
      \\"/CustomRenderFunctionSetup.vue\\": \\"/src/components/CustomRenderFunctionSetup.vue\\",
      \\"/EventCallbacksHelper.vue\\": \\"/src/components/EventCallbacksHelper.vue\\",
      \\"/ForAwaitList.vue\\": \\"/src/components/ForAwaitList.vue\\",
      \\"/SunriseSunset.vue\\": \\"/src/components/SunriseSunset.vue\\",
      \\"/grid-borders\\": {
        \\"/GridPicker.vue\\": \\"/src/components/grid-borders/GridPicker.vue\\"
      },
      \\"/style-inheritance\\": {
        \\"/ChildFour.vue\\": \\"/src/components/style-inheritance/ChildFour.vue\\",
        \\"/ChildOne.vue\\": \\"/src/components/style-inheritance/ChildOne.vue\\",
        \\"/ChildThree.vue\\": \\"/src/components/style-inheritance/ChildThree.vue\\",
        \\"/ChildTwo.vue\\": \\"/src/components/style-inheritance/ChildTwo.vue\\"
      },
      \\"/variable-fonts\\": {
        \\"/VariableFontAxis.vue\\": \\"/src/components/variable-fonts/VariableFontAxis.vue\\"
      }
    },
    \\"/use\\": {
      \\"/dataSet.ts\\": \\"/src/use/dataSet.ts\\",
      \\"/smokeTest.ts\\": \\"/src/use/smokeTest.ts\\"
    },
    \\"/views\\": {
      \\"/AboutBrowser.vue\\": \\"/src/views/AboutBrowser.vue\\",
      \\"/AccentColor.vue\\": \\"/src/views/AccentColor.vue\\",
      \\"/ConsoleTrace.vue\\": \\"/src/views/ConsoleTrace.vue\\",
      \\"/CounterExperiment.vue\\": \\"/src/views/CounterExperiment.vue\\",
      \\"/CssLayers.vue\\": \\"/src/views/CssLayers.vue\\",
      \\"/CssModules.vue\\": \\"/src/views/CssModules.vue\\",
      \\"/CssVBind.vue\\": \\"/src/views/CssVBind.vue\\",
      \\"/CssVariables.vue\\": \\"/src/views/CssVariables.vue\\",
      \\"/CustomElement.vue\\": \\"/src/views/CustomElement.vue\\",
      \\"/DialogElement.vue\\": \\"/src/views/DialogElement.vue\\",
      \\"/DynamicScopedStyles.vue\\": \\"/src/views/DynamicScopedStyles.vue\\",
      \\"/EventCallbacks.vue\\": \\"/src/views/EventCallbacks.vue\\",
      \\"/FooExperiment.vue\\": \\"/src/views/FooExperiment.vue\\",
      \\"/ForAwait.vue\\": \\"/src/views/ForAwait.vue\\",
      \\"/GridBorders.vue\\": \\"/src/views/GridBorders.vue\\",
      \\"/GridExperiment.vue\\": \\"/src/views/GridExperiment.vue\\",
      \\"/HeroiconsExperiment.vue\\": \\"/src/views/HeroiconsExperiment.vue\\",
      \\"/InputFileCapture.vue\\": \\"/src/views/InputFileCapture.vue\\",
      \\"/MultiLineSwitchStatement.vue\\": \\"/src/views/MultiLineSwitchStatement.vue\\",
      \\"/MultipleCheckboxes.vue\\": \\"/src/views/MultipleCheckboxes.vue\\",
      \\"/NthChild.vue\\": \\"/src/views/NthChild.vue\\",
      \\"/NumberFieldBinding.vue\\": \\"/src/views/NumberFieldBinding.vue\\",
      \\"/OptionalChaining.vue\\": \\"/src/views/OptionalChaining.vue\\",
      \\"/RecursionExperiment.vue\\": \\"/src/views/RecursionExperiment.vue\\",
      \\"/RenderFunction.vue\\": \\"/src/views/RenderFunction.vue\\",
      \\"/SassExperiment.vue\\": \\"/src/views/SassExperiment.vue\\",
      \\"/SelectStyling.vue\\": \\"/src/views/SelectStyling.vue\\",
      \\"/ServiceWorkerExperiment.vue\\": \\"/src/views/ServiceWorkerExperiment.vue\\",
      \\"/StyleInheritance.vue\\": \\"/src/views/StyleInheritance.vue\\",
      \\"/TaggedTemplates.vue\\": \\"/src/views/TaggedTemplates.vue\\",
      \\"/TailwindColors.vue\\": \\"/src/views/TailwindColors.vue\\",
      \\"/ThrowError.vue\\": \\"/src/views/ThrowError.vue\\",
      \\"/TimeElement.vue\\": \\"/src/views/TimeElement.vue\\",
      \\"/VariableFonts.vue\\": \\"/src/views/VariableFonts.vue\\",
      \\"/WasmExperiment.vue\\": \\"/src/views/WasmExperiment.vue\\",
      \\"/WorkerExperiment.vue\\": \\"/src/views/WorkerExperiment.vue\\",
      \\"/XorOperation.vue\\": \\"/src/views/XorOperation.vue\\"
    },
    \\"/workers\\": {
      \\"/my-first-service-worker.ts\\": \\"/src/workers/my-first-service-worker.ts\\",
      \\"/my-first-worker.ts\\": \\"/src/workers/my-first-worker.ts\\"
    }
  }
}</pre>"
`;
